### PR TITLE
Fix error handling

### PR DESCRIPTION
**Pull request by protheory8**
This is somewhat big, review carefully.

These things I added into `Cargo.toml` make all dependencies and the project itself slower to compile, but it might produce faster code (only in release mode) ([more info](https://doc.rust-lang.org/rustc/codegen-options/index.html#codegen-units)).

### CODE REVIEW

**Comment by protheory8** [comment]
Honestly, this would've been much easier if we used some crate for error handling. For your future Rust projects, you might want to check out [thiserror](https://docs.rs/thiserror/1.0.20/thiserror) crate for your libraries and [anyhow](https://docs.rs/anyhow/1.0.32/anyhow) crate for you applications. They both are created by the same author.

**Comment by protheory8** [comment]
I didn't test archive thing btw because I don't know how it works.

File: src/main.rs [hunk 1]
```diff
@@ -208,54 +208,74 @@ mod test_song_segments {
...
-	let extension = path.file_name().unwrap().to_str().unwrap().split(".").last().unwrap();
+pub fn detect_file_type(file_name: &str) -> Result<FileType, AppError> {
+	let extension = file_name.split(".").last().unwrap();
 	match extension {
-		"wav" | "ogg" | "mp3" | "flac" => FileFormat::SegmentFormat,
-		"zip" => FileFormat::SongArchiveFormat,
-		_ => FileFormat::InvalidFormat,
+		"wav" | "ogg" | "mp3" | "flac" => Ok(FileType::SegmentFormat),
+		"zip" => Ok(FileType::SongArchiveFormat),
+		_ => Err(AppError::UnrecognizedSongFormat(file_name.to_string())),
```
Comment (dyc3) [inline src/main.rs]: Instead of straight up failing when we encounter an unexpected file, we should just ignore the file.

File: src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+use std::{error::Error, fmt};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AppError {
```
Comment (dyc3) [inline src/errors.rs]:
```suggestion
pub enum DjError {
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 1 above)
Comment (protheory8) [reply src/main.rs]: It already doesn't, check the call site of `detect_file_type` (in `initialize_songs`).

Comment (dyc3) [reply src/main.rs]: Ah you're right

**Comment by protheory8** [comment]
These `.ok_or` calls are needed to convert `Option` into `Result` with this rules:
`Option::Some(val)` -> `Result::Ok(val)` where `val` is whatever is inside `Option`,
`Option::None` -> `Result::Err(e)`, where `e` is whatever we pass into `.ok_or`.

**Comment by protheory8** [comment]
`fn run(args: ArgMatches) -> Result<(), Box<dyn Error>>`
`Box<dyn Error>>` is needed so that we can return any error that implements `std::error::Error` (and not just `DjError`).

**Comment by protheory8** [comment]
I also changed `repeat_count` from `i32` to `u32`.

**Comment by protheory8** [comment]
Error message for `DjError::InvalidFileName` is kind of questionable to be honest.

**Comment by protheory8** [comment]
Btw, if you specify `max-repeats` to be 5 or lower, it panics.

**Comment by dyc3** [comment]
> Btw, if you specify `max-repeats` to be 5 or lower, it panics.

Yeah I figured. Separate issue.

**Comment by protheory8** [comment]
I left `unwrap` in a lot of places where it's not supposed to happen.
For example:
```rust
let max_repeats: u32 = args
    .value_of("max-repeats")
    .unwrap()
    .parse()
    .map_err(|_| DjError::MaxRepeatsInvalidValue)?;
```
Since `max-repeats` has a default value, `value_of` should always be `Some`, so I left `unwrap`.

**Comment by protheory8** [comment]
`.map_err` replaces one error with another one in this case.
`parse()` returns `ParseIntError` error, if we try to print it, it says something like "invalid digit found in string". Not very obvious what does it talk about, since it doesn't say anything about `max-repeats`, so I decided to replace `ParseIntError` with `DjError::MaxRepeatsInvalidValue`.

**Review by dyc3 — APPROVED** [review]

**Pull request merged**